homeassistant/components/weather/openweathermap.py

```diff
@@ -0,0 +1,160 @@
+from homeassistant.const import (
+    CONF_API_KEY, CONF_NAME, TEMP_CELSIUS, TEMP_FAHRENHEIT, CONF_LATITUDE,
+    CONF_LONGITUDE, STATE_UNKNOWN)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util import Throttle
+
+REQUIREMENTS = ['pyowm==2.4.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'OpenWeatherMap'
+ATTR_SOURCE_DETAILS = 'Data provided by OpenWeatherMap'
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=10)
+
+CONDITION_CLASSES = {
+    'cloudy': [804],
+    'fog': [701, 741],
+    'hail': [906],
+    'lightning': [210, 211, 212, 221],
+    'lightning-rainy': [200, 201, 202, 230, 231, 232],
+    'partlycloudy': [801, 802, 803],
+    'pouring': [504, 314, 502, 503, 522],
+    'rainy': [300, 301, 302, 310, 311, 312, 313, 500, 501, 520, 521],
+    'snowy': [600, 601, 602, 611, 612, 620, 621, 622],
+    'snowy-rainy': [511, 615, 616],
+    'sunny': [800],
+    'windy': [905, 951, 952, 953, 954, 955, 956, 957],
+    'windy-variant': [958, 959, 960, 961],
+    'exceptional': [711, 721, 731, 751, 761, 762, 771, 900, 901, 962, 903,
+                    904],
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Optional(CONF_LATITUDE): cv.latitude,
+    vol.Optional(CONF_LONGITUDE): cv.longitude,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the OpenWeatherMap weather platform."""
+    from pyowm import OWM
+
+    longitude = config.get(CONF_LONGITUDE, round(hass.config.longitude, 5))
+    latitude = config.get(CONF_LATITUDE, round(hass.config.latitude, 5))
+    name = config.get(CONF_NAME)
+
+    owm = OWM(config.get(CONF_API_KEY))
+
+    if not owm:
+        _LOGGER.error("Error while connecting to OpenWeatherMap")
+        return False
+
+    data = WeatherData(owm, latitude, longitude)
+
+    add_devices([OpenWeatherMapWeather(
+        name, data, hass.config.units.temperature_unit)])
+
+
+# pylint: disable=too-few-public-methods
+class OpenWeatherMapWeather(WeatherCondition):
+    """Implementation of an OpenWeatherMap sensor."""
+
+    def __init__(self, name, owm_client, temp_unit):
+        """Initialize the sensor."""
+        self._name = name
+        self.owm_client = owm_client
+        self.temp_unit = temp_unit
+        self.date = None
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def condition(self):
+        """Return the current condition."""
+        try:
+            return [k for k, v in CONDITION_CLASSES.items() if
+                    self.data.get_weather_code() in v][0]
+        except IndexError:
+            return STATE_UNKNOWN
+
+    @property
+    def temperature(self):
+        """Return the temperature."""
+        if self.temp_unit == TEMP_CELSIUS:
```

Comment:
Conversion of temperatures should happen in `WeatherEntity` base class. It cannot be done automatically because HA will only auto convert state temperatures. I've opened a similar issue for the Climate component including instructions how to work around it: https://github.com/home-assistant/home-assistant/issues/3605

Let me know if you want my assistance 👍